fearless/Modules/ChainAccountBalanceList/ChainAccountBalanceListPresenter.swift

```diff
@@ -0,0 +1,215 @@
+
+        guard let priceId = asset.priceId, let priceData = try? priceResults[priceId]?.get(),
+              let priceDecimal = Decimal(string: priceData.price) else {
+            return nil
+        }
+
+        let changeString: String = priceData.usdDayChange.map {
+            let percentValue = $0 / 100
+            return percentValue.percentString() ?? ""
+        } ?? ""
+
+        let priceString: String = usdTokenFormatterValue.stringFromDecimal(priceDecimal) ?? ""
+
+        let priceWithChangeString = [priceString, changeString].joined(separator: " ")
+
+        let priceWithChangeAttributed = NSMutableAttributedString(string: priceWithChangeString)
+
+        let color = (priceData.usdDayChange ?? 0) > 0 ? R.color.colorGreen() : R.color.colorRed()
+
+        if let color = color {
+            priceWithChangeAttributed.addAttributes(
+                [NSAttributedString.Key.foregroundColor: color],
+                range: NSRange(
+                    location: priceString.count + 1,
+                    length: changeString.count
+                )
+            )
+        }
+
+        return priceWithChangeAttributed
+    }
+
+    private func getUsdBalanceString(
+        for asset: AssetModel,
+        chain: ChainModel
+    ) -> String? {
+        let usdDisplayInfo = AssetBalanceDisplayInfo.usd()
+        let usdTokenFormatter = assetBalanceFormatterFactory.createTokenFormatter(for: usdDisplayInfo)
+        let usdTokenFormatterValue = usdTokenFormatter.value(for: selectedLocale)
+
+        let balance = getBalance(for: chain) ?? ""
+
+        guard let priceId = asset.priceId,
+              let priceData = try? priceResults[priceId]?.get(),
+              let priceDecimal = Decimal(string: priceData.price),
+              let balanceDecimal = Decimal(string: balance) else {
+            return nil
+        }
+
+        let totalBalanceDecimal = priceDecimal * balanceDecimal
+
+        return usdTokenFormatterValue.stringFromDecimal(totalBalanceDecimal)
+    }
+
+    private func provideViewModel() {
+        let usdDisplayInfo = AssetBalanceDisplayInfo.usd()
+        let usdTokenFormatter = assetBalanceFormatterFactory.createTokenFormatter(for: usdDisplayInfo)
+        let usdTokenFormatterValue = usdTokenFormatter.value(for: selectedLocale)
+
+        var totalWalletBalance: Decimal = chainModels.compactMap { chainModel in
+
+            chainModel.assets.compactMap {
+                let balance = getBalance(for: chainModel) ?? ""
+
+                guard let priceId = $0.priceId,
+                      let priceData = try? priceResults[priceId]?.get(),
+                      let priceDecimal = Decimal(string: priceData.price),
+                      let balanceDecimal = Decimal(string: balance) else {
+                    return nil
+                }
+
+                return priceDecimal * balanceDecimal
+            }.reduce(0, +)
+        }.reduce(0, +)
+
+        viewModels = chainModels.map { chainModel in
```

Comment:
maybe instead `compactMap { compactMap { ... } }` do `.reduce([], { $0 + $1.chainModel.assets }.compactMap { ... }` with merged array? chain-way is preferred

Reply:
@posplaw the chainModel argument from first compactMap is used inside block. just array of assets not enough to build necessary objects